ARROW-12679: [Java] JDBC->Arrow for NOT NULL columns.

jmgpeeters opened the pull request:
Prior to this patch, the VectorSchemaRoot's schema coming out of the JDBC adaptor always has all columns as nullable, even when the SQL column is NOT NULL. Even if this has no immediate impact on performance, it throws away information that can be useful downstream.

The fix just replaces the `true` (for nullable) by the actual `isNullableColumn` information.

liyafan82 commented on java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowOptionalColumnsTest.java:
```diff
@@ -0,0 +1,91 @@
+
+/**
+ * JUnit Test Class which contains methods to test JDBC to Arrow data conversion functionality for
+ * (non-)optional columns, in particular with regard to the ensuing VectorSchemaRoot's schema.
```
ensuing -> ensuring?

jmgpeeters replied:
ensuing: "occurring afterwards or as a result." (as per google) - a fairly rare word perhaps, but I meant the "as a result" meaning. :)

liyafan82 replied:
Thanks for the explanation. I have learned a new word :)

liyafan82 commented on java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowOptionalColumnsTest.java:
```diff
@@ -0,0 +1,91 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
Why do we start a new test class?
Can we place the tests in `JdbcToArrowDataTypesTest`?

jmgpeeters replied:
Sure, we could. As this wasn't really about testing int, float, string, ... etc, though, I thought it was clearer to have a small test class specifically for verifying the schema. No strong feelings, I can refactor.

liyafan82 replied:
I have no strong feelings either. So you can make the decision.

liyafan82 commented on java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowOptionalColumnsTest.java:
```diff
@@ -0,0 +1,91 @@
+  };
+
+  /**
+   * Constructor which populate table object for each test iteration.
```
nit: populate -> populates?

jmgpeeters replied:
Yep. I'm pretty sure I copy-pasted that comment from elsewhere, so will fix throughout.

liyafan82 commented on java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowOptionalColumnsTest.java:
```diff
@@ -0,0 +1,91 @@
+   */
+  @Parameterized.Parameters
+  public static Collection<Object[]> getTestData() throws SQLException, ClassNotFoundException, IOException {
+    return Arrays.asList(prepareTestData(testFiles, JdbcToArrowNullTest.class));
```
should be `JdbcToArrowOptionalColumnsTest.class`?

jmgpeeters replied:
Yes.

liyafan82 commented on java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowOptionalColumnsTest.java:
```diff
@@ -0,0 +1,91 @@
+  @Test
+  public void testJdbcToArrowValues() throws SQLException, IOException {
+    testDataSets(JdbcToArrow.sqlToArrow(conn, table.getQuery(), new RootAllocator(Integer.MAX_VALUE)));
+  }
```
We need to close the allocator somewhere?

jmgpeeters replied:
I suppose we do, yes, but I built this test based on existing tests, which don't seem to close the allocator either, see e.g. JdbcToArrowNullTest and JdbcToArrowDataTypesTest. 

The memory is probably small enough, and the process short-lived enough, that it doesn't matter.

liyafan82 commented on java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowOptionalColumnsTest.java:
```diff
@@ -0,0 +1,91 @@
+    JdbcToArrowTestHelper.assertFieldMetadataIsEmpty(root);
+
+    assertTrue(root.getSchema().getFields().get(0).isNullable());
+    assertFalse(root.getSchema().getFields().get(1).isNullable());
```
We also need a case for which the field is non-nullable?

jmgpeeters commented:
@liyafan82 I've addressed the comments. 
- I left the new test class in place - I think it's slightly clearer this way but indeed not so important,
- I didn't change anything to the closing of the allocator, as it's the same pattern that's used in the other tests as well - and I think it's a reasonable one here as this is short-lived and low-footprint memory. If we'd change it here, we should change it everywhere (which I'd be happy to do).

liyafan82 reviewed (approved):
LGTM. Will merge soon if there are no more comments.
